[rustdoc] If re-export is private, get the next item until a public one is found or expose the private item directly

GuillaumeGomez opened the pull request:
Fixes #81141.

If we have:

```rust
use Private as Something;

pub fn foo() -> Something {}
```

Then `Something` will be replaced by `Private`.

r? @notriddle

GuillaumeGomez committed:
If re-export is private, get the next item until a public one is found or expose the private item directly

GuillaumeGomez committed:
Add regression test for #81141

fmease commented:
I haven't looked too much at the implementation yet but do you think it'd be worthwhile to kick off a perf run since it looks like in the worst case this is running a nested loop with a nesting level of up to 4 (!) whenever it cleans a hir qpath.

GuillaumeGomez commented:
Good point!

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 573a7f4ee75cbd7640143e6687d9d06cec1d7062 with merge cfeb9c6f60982932992f2550ba9b986fcc4e9ef7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"573a7f4ee75cbd7640143e6687d9d06cec1d7062","merge_sha":"cfeb9c6f60982932992f2550ba9b986fcc4e9ef7"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5469209492/jobs/9958956419)
Build commit: cfeb9c6f60982932992f2550ba9b986fcc4e9ef7 (`cfeb9c6f60982932992f2550ba9b986fcc4e9ef7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5469209492/jobs/9958956419"},"merge_sha":"cfeb9c6f60982932992f2550ba9b986fcc4e9ef7"} -->

rust-timer commented:
Queued cfeb9c6f60982932992f2550ba9b986fcc4e9ef7 with parent 5dac6b320be868f898a3c753934eabc79ff2e406, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (cfeb9c6f60982932992f2550ba9b986fcc4e9ef7): [comparison URL](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.9%, 0.9%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.9%, 0.9%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [1.7%, 2.8%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-2.1%, -1.1%] | 2     |
| All ❌✅ (primary)                 | 0.4%  | [0.4%, 0.4%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=5dac6b320be868f898a3c753934eabc79ff2e406&end=cfeb9c6f60982932992f2550ba9b986fcc4e9ef7&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 658.838s -> 658.176s (-0.10%)

GuillaumeGomez commented:
Seems like the regression isn't too big. At least way less than what I originally feared.

fmease commented:
Yay, seems like it! Or our benchmark & stress test suite is not big enough :P

I'm gonna go over the changes tonight or tomorrow hopefully if I find the time.

I wonder if there is any rustc API in `rustc_resolve` or anything from intra-doc link module that can we could leverage but it doesn't seem so, dunno

GuillaumeGomez commented:
The problem here is trying to retrieve the `DefId` and then the HIR element again. Back and forth fun.

> Yay, seems like it! Or our benchmark & stress test suite is not big enough :P

It's a possibility indeed! :)

> I'm gonna go over the changes tonight or tomorrow hopefully if I find the time.

Thanks!

notriddle reviewed (changes requested):
Is there no facility to accomplish this already in the compiler? I know it suggests paths to import, which means it *should* try to avoid giving inaccessible ones.

If not, here's a suggestion for fixing (and more thoroughly testing) this function.

GuillaumeGomez commented:
> Is there no facility to accomplish this already in the compiler? I know it suggests paths to import, which means it should try to avoid giving inaccessible ones.

I didn't find them at least. Doesn't mean they don't exist, but I missed them if they do.

GuillaumeGomez committed:
Improve code readability

GuillaumeGomez committed:
Extend issue-81141-private-reexport-in-public-api test to cover more cases

GuillaumeGomez commented:
I applied your suggestions @notriddle, thanks a lot!

notriddle reviewed (approved):
Looks good. Awaiting fmease’s feedback

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1480,8 +1480,89 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+                    let Some(local_use_def_id) = use_def_id.as_local()
+                {
+                    let hir = cx.tcx.hir();
+                    // let parent_mod = hir.local_def_id_to_hir_id();
```
```suggestion
```

GuillaumeGomez replied:
Outch good catch! Removed it.

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1480,8 +1480,88 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+    for child in
+        cx.tcx.module_children_local(parent_def_id).iter().filter(move |c| c.ident == ident)
+    {
+        if let Res::Def(DefKind::Ctor(..), _) | Res::SelfCtor(..) = child.res {
```
Of course, we could at some point support `Ctor` but that's low priority. This would only be reachable with const generics (and const expressions if/once we render their body). E.g. `Container<{ Alias }>` with `use Type::Ctor as Alias`.

GuillaumeGomez replied:
For now it's something we exclude everywhere in rustdoc when looking for reexports. We can indeed add support for it later on.

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1480,8 +1480,88 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+                                if let Res::Def(DefKind::Ctor(..), _) | Res::SelfCtor(..) = res {
+                                    continue;
+                                }
+                                if !cx.tcx.is_doc_hidden(use_def_id) &&
```
This doesn't respect `--document-private-items` and `--document-hidden-items`. Not sure if what the desired behavior would be though. Do we want to keep showing the private or hidden alias when such flag is passed? Not blocking.

GuillaumeGomez replied:
I was planning to actually fix it in a follow-up (all set-up in my calendar and everything :D).

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1480,8 +1480,88 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+/// or `doc(hidden)`). If it's not possible, it'll return the "end type".
+///
+/// If the path is not a re-export or is public, it'll return `None`.
+fn first_not_private(
```
Minor, sounds better (I'm not a native speaker tho).

```suggestion
fn first_non_private(
```

GuillaumeGomez replied:
Sounds better indeed, let's go with it!

fmease reviewed (approved):
Let's say I'm not super familiar with this area of the code (fetching things from HIR and how re-exports are represented) but it looks reasonable.

GuillaumeGomez committed:
Rename `first_not_private` into `first_non_private`

GuillaumeGomez commented:
Updated the function name.

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1479,8 +1479,88 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+        [] => return None,
+        // Relative paths are available in the same scope as the owner.
+        [leaf] => (cx.tcx.local_parent(hir_id.owner.def_id), leaf.ident),
+        // So are self paths.
+        [parent, leaf] if parent.ident.name == kw::SelfLower => {
+            (cx.tcx.local_parent(hir_id.owner.def_id), leaf.ident)
+        }
+        // Crate paths are not. We start from the crate root.
+        [parent, leaf] if parent.ident.name == kw::Crate => {
+            (LOCAL_CRATE.as_def_id().as_local()?, leaf.ident)
+        }
```
This also needs to handle `kw::Super`.

fmease replied:
Consider:

```rs
use external::Public as Private;

pub mod external {
   pub struct Public;

   pub fn make() -> super::Private { super::Private }
   //pub fn make() -> crate::Private { crate::Private }
}
```

GuillaumeGomez replied:
Good point. I'll add a test as well, good catch!

fmease replied:
As well as `kw::PathRoot` (`{{root}}`) for `::Private` in edition 2015.

```rs
// edition:2015
pub fn make() -> ::Private { ::Private }
```

GuillaumeGomez replied:
I think we can specify the edition we want in tests so should be simple.

GuillaumeGomez committed:
Correctly handle `super` and `::`

GuillaumeGomez commented:
Added the handling for `super` and for `::` (2015 edition).

fmease reviewed (approved):
Technically speaking, `edition:2015` isn't necessary because it's the default in tests but I like that it's there since it makes it explicit to the reader.

It makes me a bit uneasy that we straight up interpret `kw::PathRoot` as `kw::Crate` in any old edition instead of checking `parent.ident.span.is_rust_2015()` first (if that's actually how you properly check this), although I'm not sure if it can cause any problems right now. For example in Rust 2018+, `::dep` always refers to a *crate* but it's unclear to me if in such case `fn first_non_private` is actually reached (e.g. I've quickly tried `pub use ::dep as alias;` but that didn't lead to `first_non_private` being called).

Edit: I think we are good to go. Thanks for working on it and for considering my feedback! :)
Edit: (Typo in commit message: *Correctly **handly** `super` and `::`*)

GuillaumeGomez commented:
If you can find a case where the `::` handling is invalid, don't hesitate to share it or to open a PR to fix it (very curious to see in which case it could have a different meaning).

> Edit: I think we are good to go. Thanks for working on it and for considering my feedback! :)

The thanks is not supposed to work this way for the other! You're the one who provided great feedback and made a great review. I'm the one who is (as should be) thankful!

> Edit: (Typo in commit message: *Correctly __handly__ `super` and `::`*)

Mouhahaha (fixed).

Anyway, thanks a lot to both of you for the great reviews!

@bors r=notriddle,fmease rollup

bors commented:
:pushpin: Commit 4a1e06b453d14b61d4fc19532918694f485bf51e has been approved by `notriddle,fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,fmease 4a1e06b453d14b61d4fc19532918694f485bf51e -->
<!-- homu: {"type":"Approved","sha":"4a1e06b453d14b61d4fc19532918694f485bf51e","approver":"notriddle,fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- rollup=maybe this seems to have caused problems in above rollups when documenting the compiler

GuillaumeGomez commented:
Just ran `./x.py doc compiler/ --stage 1` locally and no bug or anything. I rebased on `master` just in case. Let's keep it out of rollups for the time being. If it fails again, I'll try to check what's different between bors build and local one.

@bors r=notriddle,fmease rollup=iffy

bors commented:
:pushpin: Commit 3c75113cd920ea547c8db8fdd07893995b27e45c has been approved by `notriddle,fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,fmease 3c75113cd920ea547c8db8fdd07893995b27e45c -->
<!-- homu: {"type":"Approved","sha":"3c75113cd920ea547c8db8fdd07893995b27e45c","approver":"notriddle,fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3c75113cd920ea547c8db8fdd07893995b27e45c with merge b73c86f09536c9a3ef3ccc81bc23e8ce63985ca2...
<!-- homu: {"type":"BuildStarted","head_sha":"3c75113cd920ea547c8db8fdd07893995b27e45c","merge_sha":"b73c86f09536c9a3ef3ccc81bc23e8ce63985ca2"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5501898187/jobs/10026059176)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5501898187/jobs/10026059176","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5501898187/jobs/10025799634) [(plain)](https://github.com/rust-lang-ci/rust/commit/b73c86f09536c9a3ef3ccc81bc23e8ce63985ca2/checks/14894023500/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting rustc_monomorphize v0.0.0 (/checkout/compiler/rustc_monomorphize)
 Documenting rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
 Documenting rustc_symbol_mangling v0.0.0 (/checkout/compiler/rustc_symbol_mangling)
 Documenting rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
thread 'rustc' panicked at 'assertion failed: cx.impl_trait_bounds.is_empty()', src/librustdoc/clean/utils.rs:542:5
   0:     0xffffa0cf8d50 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hdecd0b6e83797efc
   1:     0xffffa0d5f820 - core::fmt::write::ha2a3999375bbe080
   2:     0xffffa0cee294 - std::io::Write::write_fmt::hb374751c6d88c886
   3:     0xffffa0cf8ba0 - std::sys_common::backtrace::print::hb4051d836ca66a2f
---
   8:     0xffffa0cfc030 - std::panicking::begin_panic_handler::{{closure}}::h33cee63097d4ff20
   9:     0xffffa0cf91cc - std::sys_common::backtrace::__rust_end_short_backtrace::h39eb16a549ed1bfe
  10:     0xffffa0cfbda4 - rust_begin_unwind
  11:     0xffffa0cbe110 - core::panicking::panic_fmt::h311203a6d22a4446
  12:     0xffffa0cbe180 - core::panicking::panic::h940cf507e8cfdda6
  13:     0xaaaac4270880 - rustdoc[ab442b94caf5758e]::clean::utils::enter_impl_trait::<rustdoc[ab442b94caf5758e]::clean::clean_function::{closure#0}, (rustdoc[ab442b94caf5758e]::clean::types::Generics, rustdoc[ab442b94caf5758e]::clean::types::FnDecl)>
  14:     0xaaaac448a868 - rustdoc[ab442b94caf5758e]::clean::clean_fn_or_proc_macro
  15:     0xaaaac4401d58 - <rustdoc[ab442b94caf5758e]::core::DocContext>::with_param_env::<alloc[4b03aabee438587d]::vec::Vec<rustdoc[ab442b94caf5758e]::clean::types::Item>, rustdoc[ab442b94caf5758e]::clean::clean_maybe_renamed_item::{closure#1}>
  16:     0xaaaac447f4d0 - <&mut rustdoc[ab442b94caf5758e]::clean::clean_doc_module::{closure#2} as core[ded48306656406af]::ops::function::FnOnce<(&(&rustc_hir[14dc881e803d143a]::hir::Item, core[ded48306656406af]::option::Option<rustc_span[1a9050cee52d4ca2]::symbol::Symbol>, core[ded48306656406af]::option::Option<rustc_span[1a9050cee52d4ca2]::def_id::LocalDefId>),)>>::call_once
  17:     0xaaaac437d460 - <alloc[4b03aabee438587d]::vec::Vec<rustdoc[ab442b94caf5758e]::clean::types::Item> as alloc[4b03aabee438587d]::vec::spec_extend::SpecExtend<rustdoc[ab442b94caf5758e]::clean::types::Item, core[ded48306656406af]::iter::adapters::flatten::FlatMap<indexmap[7b105bfaced6789e]::map::iter::Values<(rustc_span[1a9050cee52d4ca2]::def_id::LocalDefId, core[ded48306656406af]::option::Option<rustc_span[1a9050cee52d4ca2]::symbol::Symbol>), (&rustc_hir[14dc881e803d143a]::hir::Item, core[ded48306656406af]::option::Option<rustc_span[1a9050cee52d4ca2]::symbol::Symbol>, core[ded48306656406af]::option::Option<rustc_span[1a9050cee52d4ca2]::def_id::LocalDefId>)>, alloc[4b03aabee438587d]::vec::Vec<rustdoc[ab442b94caf5758e]::clean::types::Item>, rustdoc[ab442b94caf5758e]::clean::clean_doc_module::{closure#2}>>>::spec_extend
  18:     0xaaaac4484b58 - rustdoc[ab442b94caf5758e]::clean::clean_doc_module
  19:     0xaaaac447ee44 - <&mut rustdoc[ab442b94caf5758e]::clean::clean_doc_module::{closure#1} as core[ded48306656406af]::ops::function::FnMut<(&rustdoc[ab442b94caf5758e]::visit_ast::Module,)>>::call_mut
  20:     0xaaaac437c9c8 - <alloc[4b03aabee438587d]::vec::Vec<rustdoc[ab442b94caf5758e]::clean::types::Item> as alloc[4b03aabee438587d]::vec::spec_extend::SpecExtend<rustdoc[ab442b94caf5758e]::clean::types::Item, core[ded48306656406af]::iter::adapters::filter_map::FilterMap<core[ded48306656406af]::slice::iter::Iter<rustdoc[ab442b94caf5758e]::visit_ast::Module>, rustdoc[ab442b94caf5758e]::clean::clean_doc_module::{closure#1}>>>::spec_extend
  21:     0xaaaac4484b18 - rustdoc[ab442b94caf5758e]::clean::clean_doc_module
  22:     0xaaaac426d014 - rustdoc[ab442b94caf5758e]::clean::utils::krate
  23:     0xaaaac429b0d8 - <rustc_session[ff0f14629b323e96]::session::Session>::time::<rustdoc[ab442b94caf5758e]::clean::types::Crate, rustdoc[ab442b94caf5758e]::core::run_global_ctxt::{closure#5}>
  24:     0xaaaac4404e98 - rustdoc[ab442b94caf5758e]::core::run_global_ctxt
  25:     0xaaaac429b5d4 - <rustc_session[ff0f14629b323e96]::session::Session>::time::<(rustdoc[ab442b94caf5758e]::clean::types::Crate, rustdoc[ab442b94caf5758e]::config::RenderOptions, rustdoc[ab442b94caf5758e]::formats::cache::Cache), rustdoc[ab442b94caf5758e]::main_args::{closure#1}::{closure#0}::{closure#0}::{closure#0}>
  26:     0xaaaac4415f38 - <rustc_middle[fc036cd1a73b79fc]::ty::context::GlobalCtxt>::enter::<rustdoc[ab442b94caf5758e]::main_args::{closure#1}::{closure#0}::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>
  27:     0xaaaac44a7ca8 - <rustc_interface[5ee08df1f5122acb]::interface::Compiler>::enter::<rustdoc[ab442b94caf5758e]::main_args::{closure#1}::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>
  28:     0xaaaac43358ec - rustc_span[1a9050cee52d4ca2]::set_source_map::<core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>, rustc_interface[5ee08df1f5122acb]::interface::run_compiler<core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>, rustdoc[ab442b94caf5758e]::main_args::{closure#1}>::{closure#0}::{closure#0}>
  29:     0xaaaac4304638 - <scoped_tls[82dde1795b3d2521]::ScopedKey<rustc_span[1a9050cee52d4ca2]::SessionGlobals>>::set::<rustc_interface[5ee08df1f5122acb]::interface::run_compiler<core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>, rustdoc[ab442b94caf5758e]::main_args::{closure#1}>::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>
  30:     0xaaaac4432c84 - std[33fb26fdd69d5d3a]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[5ee08df1f5122acb]::util::run_in_thread_pool_with_globals<rustc_interface[5ee08df1f5122acb]::interface::run_compiler<core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>, rustdoc[ab442b94caf5758e]::main_args::{closure#1}>::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>
  31:     0xaaaac4479d54 - <<std[33fb26fdd69d5d3a]::thread::Builder>::spawn_unchecked_<rustc_interface[5ee08df1f5122acb]::util::run_in_thread_pool_with_globals<rustc_interface[5ee08df1f5122acb]::interface::run_compiler<core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>, rustdoc[ab442b94caf5758e]::main_args::{closure#1}>::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ded48306656406af]::result::Result<(), rustc_span[1a9050cee52d4ca2]::ErrorGuaranteed>>::{closure#1} as core[ded48306656406af]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  33:     0xffffa0c1a624 - start_thread
  34:     0xffffa0b4d49c - <unknown>
  35:                0x0 - <unknown>


error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-rustdoc&template=ice.md
note: rustc 1.73.0-nightly (b73c86f09 2023-07-09) running on aarch64-unknown-linux-gnu


note: compiler flags: --crate-type lib -Z unstable-options -Z unstable-options -C symbol-mangling-version=v0 -Z unstable-options -Z unstable-options -Z normalize-docs -Z crate-attr=warn(rust_2018_idioms) -Z force-unstable-if-unmarked -Z unstable-options
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
end of query stack
error: could not document `rustc_trait_selection`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_trait_selection compiler/rustc_trait_selection/src/lib.rs --target aarch64-unknown-linux-gnu -o /checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'values(feature)' --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=2af3d3565ab26afd -L dependency=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/release/deps --extern rustc_ast=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_ast-583dc69f4da5bf5f.rmeta --extern rustc_attr=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_attr-c94138757b643f34.rmeta --extern rustc_data_structures=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_data_structures-00a60c154003094a.rmeta --extern rustc_errors=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_errors-8380795c15d2331e.rmeta --extern rustc_fluent_macro=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/release/deps/librustc_fluent_macro-5ece8e1e21b27d23.so --extern rustc_hir=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_hir-55258cf235147450.rmeta --extern rustc_index=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_index-9f93f14d714390fb.rmeta --extern rustc_infer=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_infer-2d432b800a6c795a.rmeta --extern rustc_macros=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/release/deps/librustc_macros-464628d5e62c04e5.so --extern rustc_middle=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_middle-53ac2c9bab58830e.rmeta --extern rustc_parse_format=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_parse_format-5d7099e53dacc36e.rmeta --extern rustc_query_system=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_query_system-a98704144a54db62.rmeta --extern rustc_serialize=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_serialize-b3c8cf6764c871d4.rmeta --extern rustc_session=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_session-e7c65f2dfda30378.rmeta --extern rustc_span=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_span-bd7a05aa7c15e9af.rmeta --extern rustc_target=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_target-d65de9ddf89d0d34.rmeta --extern rustc_transmute=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_transmute-f6ecfe3f6891b93e.rmeta --extern smallvec=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/libsmallvec-ed8b7cc5df53142c.rmeta --extern tracing=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-rustc/aarch64-unknown-linux-gnu/release/deps/libtracing-1e48c348890d7c9e.rmeta --extern-html-root-url 'smallvec=https://docs.rs/smallvec/1.10.0/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.37/' -Zunstable-options --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.73.0-nightly (b73c86f09 2023-07-09)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition '-Zcrate-attr=warn(rust_2018_idioms)' --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 101)
[RUSTC-TIMING] syn test:false 7.582
Build completed unsuccessfully in 0:35:07
```

</details>

GuillaumeGomez commented:
Is there a difference compared to running `x.py doc compiler --stage 1`?

GuillaumeGomez committed:
Add test for private items

GuillaumeGomez committed:
Add support for `--document-hidden-items` in `first_non_private`

GuillaumeGomez committed:
Add test for `--document-hidden-items`

GuillaumeGomez committed:
Re-add missing generics in `first_non_private`

GuillaumeGomez committed:
Add regression test for generics reexport of private import

GuillaumeGomez commented:
I found the bug: we needed to put back the generics into the `Path`. They are "lost" because there aren't generics in `use` statements.

I also added support for `--document-hidden-items`. About `--document-private-items`: since only public re-exports are displayed in the documentation, there was nothing to do. I added a test for both in any case.

GuillaumeGomez commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1479,8 +1479,121 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+                    // we add them by hand.
+                    let mut segments = new_path.segments.to_vec();
+                    if let Some(last) = segments.last_mut() {
+                        // `transmute` is needed because we are using a wrong lifetime. Since
+                        // `segments` will be dropped at the end of this block, it's fine.
```
I did that originally but I wasn't happy about potentially cleaning the generics twice. However, the code looks much better this way so let's do it like this. Thanks!

GuillaumeGomez committed:
Remove needs for transmute

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit ea9a17b9995b7a076283777b7d462a360fece2d6 with merge f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ea9a17b9995b7a076283777b7d462a360fece2d6","merge_sha":"f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5520603799/jobs/10069578746)
Build commit: f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b (`f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5520603799/jobs/10069578746"},"merge_sha":"f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b"} -->

rust-timer commented:
Queued f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b with parent 63ef74b6aa0448cee884324a1f47dde284176895, future [comparison URL](https://perf.rust-lang.org/compare.html?start=63ef74b6aa0448cee884324a1f47dde284176895&end=f0d752f9734b7fb9629efcfc8e5c3861d41b8c5b).
<!-- rust-timer: temporary -->

GuillaumeGomez commented:
Oof. That's way too massive. I'm surprised there is such a difference...

GuillaumeGomez committed:
Revert "Remove needs for transmute"

This reverts commit ea9a17b9995b7a076283777b7d462a360fece2d6.

GuillaumeGomez commented:
I rebased on top of master and only reverted the last commit to see if the regression comes from there.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c06ae43e33a8faae1e5d17acf882b18e937b35e4 with merge 8eec6982e7c85b780374a432255b8a5de7962662...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c06ae43e33a8faae1e5d17acf882b18e937b35e4","merge_sha":"8eec6982e7c85b780374a432255b8a5de7962662"} -->

rust-timer commented:
Queued 8eec6982e7c85b780374a432255b8a5de7962662 with parent 33a2c2487ac5d9927830ea4c1844335c6b9f77db, future [comparison URL](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (8eec6982e7c85b780374a432255b8a5de7962662): [comparison URL](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 9.7%  | [0.6%, 18.9%]  | 2     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.4%, 0.4%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.3%, -0.4%] | 13    |
| Improvements ✅ <br /> (secondary) | -0.7% | [-1.0%, -0.4%] | 6     |
| All ❌✅ (primary)                 | 0.5%  | [-1.3%, 18.9%] | 15    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 10.8% | [10.8%, 10.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.1% | [-5.1%, -3.6%] | 6     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.0% | [-5.1%, 10.8%] | 7     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=33a2c2487ac5d9927830ea4c1844335c6b9f77db&end=8eec6982e7c85b780374a432255b8a5de7962662&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 658.224s -> 657.583s (-0.10%)

GuillaumeGomez committed:
Cache qpath first public result

GuillaumeGomez commented:
Let's see the caching changes.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 54a333f67d94fa5012c83dff6dd88ccd5bd7acb6 with merge 6f538cef25d159358a67cfcdbb5618b0c91c2114...
<!-- homu: {"type":"TryBuildStarted","head_sha":"54a333f67d94fa5012c83dff6dd88ccd5bd7acb6","merge_sha":"6f538cef25d159358a67cfcdbb5618b0c91c2114"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5611438402/job/15205622824)
Build commit: 6f538cef25d159358a67cfcdbb5618b0c91c2114 (`6f538cef25d159358a67cfcdbb5618b0c91c2114`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5611438402/job/15205622824"},"merge_sha":"6f538cef25d159358a67cfcdbb5618b0c91c2114"} -->

rust-timer commented:
Queued 6f538cef25d159358a67cfcdbb5618b0c91c2114 with parent b14fd2359f47fb9a14bbfe55359db4bb3af11861, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114).
There is currently 1 other artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (6f538cef25d159358a67cfcdbb5618b0c91c2114): [comparison URL](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 9.1%  | [0.4%, 17.8%]  | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 1     |
| All ❌✅ (primary)                 | 9.1%  | [0.4%, 17.8%]  | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9%  | [1.9%, 1.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.6%  | [2.6%, 2.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.9%, -2.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-2.9%, 1.9%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.3% | [1.4%, 9.2%] | 8     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.3% | [1.4%, 9.2%] | 8     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b14fd2359f47fb9a14bbfe55359db4bb3af11861&end=6f538cef25d159358a67cfcdbb5618b0c91c2114&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 649.458s -> 647.46s (-0.31%)

GuillaumeGomez commented:
So caching with `HirId` is not good enough apparently.

GuillaumeGomez committed:
Improve performance of `first_non_private`

GuillaumeGomez commented:
I realized that an entire loop could be removed (silly me not remembering about `find_by_def_id`...).

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1f828f01552b54abb8e33610484d9269067fcc28 with merge e55147822ced7759651867efebb824fff4c64a43...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1f828f01552b54abb8e33610484d9269067fcc28","merge_sha":"e55147822ced7759651867efebb824fff4c64a43"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5657566518/job/15329642608)
Build commit: e55147822ced7759651867efebb824fff4c64a43 (`e55147822ced7759651867efebb824fff4c64a43`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5657566518/job/15329642608"},"merge_sha":"e55147822ced7759651867efebb824fff4c64a43"} -->

rust-timer commented:
Queued e55147822ced7759651867efebb824fff4c64a43 with parent ff8fe76c0e36b65c038a080a8a8341024104d117, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43).
There are currently 5 other artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~7.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e55147822ced7759651867efebb824fff4c64a43): [comparison URL](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression



**Warning ⚠**: The following benchmark(s) failed to build:
- match-stress


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [0.6%, 2.3%] | 4     |
| Regressions ❌ <br /> (secondary)  | 0.5% | [0.5%, 0.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [0.6%, 2.3%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-2.8%, -2.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.3%, -1.3%] | 2     |
| All ❌✅ (primary)                 | -2.5% | [-2.8%, -2.3%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=ff8fe76c0e36b65c038a080a8a8341024104d117&end=e55147822ced7759651867efebb824fff4c64a43&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 651.831s -> 649.821s (-0.31%)

GuillaumeGomez commented:
Seems like I fixed most of the perf regression! 🎉 

Just need to fix the newly introduced bug and it should be good to go. :)

fmease commented:
You could get rid of `transmute` altogether by using `tcx.arena.alloc_slice`. Not sure if that has any perf implication tho.
Check out:
https://github.com/fmease/rust/pull/8/commits/03a98899b5317306c4c01b1a84e78d7aa7e93b37

GuillaumeGomez committed:
Fix regression for private in public

GuillaumeGomez commented:
Fixed the bug (needed to check that there is actually an ident on the `use`), added the regression test. The perf impact has been greatly reduced. I think it's now ready to go. :)

GuillaumeGomez commented:
Forgot to `git add` the regression test. Added it.

GuillaumeGomez commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1496,8 +1496,138 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
+    if let Some(last) = new_path_segments.last_mut() {
+        // `transmute` is needed because we are using a wrong lifetime. Since
+        // `segments` will be dropped at the end of this block, it's fine.
+        last.args = unsafe { transmute(path.segments.last().as_ref().unwrap().args.clone()) };
```
I'll cherry-pick @fmease suggestion then.

GuillaumeGomez commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 531b8ccd83359282f12e4071418bdd55e0d5b7ba with merge 4bc991ec73f6c11ef8c32bbc679a6989e07cbd91...
<!-- homu: {"type":"TryBuildStarted","head_sha":"531b8ccd83359282f12e4071418bdd55e0d5b7ba","merge_sha":"4bc991ec73f6c11ef8c32bbc679a6989e07cbd91"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5672489277/job/15373845515)
Build commit: 4bc991ec73f6c11ef8c32bbc679a6989e07cbd91 (`4bc991ec73f6c11ef8c32bbc679a6989e07cbd91`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5672489277/job/15373845515"},"merge_sha":"4bc991ec73f6c11ef8c32bbc679a6989e07cbd91"} -->

rust-timer commented:
Queued 4bc991ec73f6c11ef8c32bbc679a6989e07cbd91 with parent 601a34de8c10458b72a7781eb0b44a7981e4a2b1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91).
There are currently 3 other artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~4.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4bc991ec73f6c11ef8c32bbc679a6989e07cbd91): [comparison URL](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [2.6%, 2.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | 2.6%  | [2.6%, 2.6%]   | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-3.5%, -3.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.5% | [2.5%, 2.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.5% | [2.5%, 2.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=601a34de8c10458b72a7781eb0b44a7981e4a2b1&end=4bc991ec73f6c11ef8c32bbc679a6989e07cbd91&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.146s -> 649.937s (-0.03%)

GuillaumeGomez committed:
Remove transmute calls and caching for use paths

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 2461d0cf9cca7e437e17cc251496963caeac051d with merge 5224c258f207a3f5b4c32980f776a9317118ef1e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2461d0cf9cca7e437e17cc251496963caeac051d","merge_sha":"5224c258f207a3f5b4c32980f776a9317118ef1e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5678554039/job/15391163315)
Build commit: 5224c258f207a3f5b4c32980f776a9317118ef1e (`5224c258f207a3f5b4c32980f776a9317118ef1e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5678554039/job/15391163315"},"merge_sha":"5224c258f207a3f5b4c32980f776a9317118ef1e"} -->

rust-timer commented:
Queued 5224c258f207a3f5b4c32980f776a9317118ef1e with parent 3e11b223d7ad9836babf2d531b0ca6d13ff3f065, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3e11b223d7ad9836babf2d531b0ca6d13ff3f065&end=5224c258f207a3f5b4c32980f776a9317118ef1e).
There are currently 7 other artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~9.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

GuillaumeGomez commented:
So to sum things up:

| method | mean | range |
| - | - | - |
| transmute + cache | 1.1% | [0.6%, 2.3%] |
| arena + cache | 2.6% | [2.6%, 2.6%] |
| early clean | 1.5% | [0.3%, 2.7%] |

I think the current result is good enough. What do you think @notriddle ?

bors commented:
:pushpin: Commit 2461d0cf9cca7e437e17cc251496963caeac051d has been approved by `notriddle,fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,fmease 2461d0cf9cca7e437e17cc251496963caeac051d -->
<!-- homu: {"type":"Approved","sha":"2461d0cf9cca7e437e17cc251496963caeac051d","approver":"notriddle,fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
@bors rollup=never

bors commented:
:hourglass: Testing commit 2461d0cf9cca7e437e17cc251496963caeac051d with merge 9339f446a5302cd5041d3f3b5e59761f36699167...
<!-- homu: {"type":"BuildStarted","head_sha":"2461d0cf9cca7e437e17cc251496963caeac051d","merge_sha":"9339f446a5302cd5041d3f3b5e59761f36699167"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5682771582/job/15405044668)
Approved by: notriddle,fmease
Pushing 9339f446a5302cd5041d3f3b5e59761f36699167 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"notriddle,fmease","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5682771582/job/15405044668"},"merge_sha":"9339f446a5302cd5041d3f3b5e59761f36699167"} -->

rust-timer commented:
Finished benchmarking commit (9339f446a5302cd5041d3f3b5e59761f36699167): [comparison URL](https://perf.rust-lang.org/compare.html?start=b73e9a48aeeb44fa897ab820737dfd77c5076e01&end=9339f446a5302cd5041d3f3b5e59761f36699167&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b73e9a48aeeb44fa897ab820737dfd77c5076e01&end=9339f446a5302cd5041d3f3b5e59761f36699167&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0% | [0.3%, 2.7%] | 4     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.3%, 0.5%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.0% | [0.3%, 2.7%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b73e9a48aeeb44fa897ab820737dfd77c5076e01&end=9339f446a5302cd5041d3f3b5e59761f36699167&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [2.2%, 2.6%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b73e9a48aeeb44fa897ab820737dfd77c5076e01&end=9339f446a5302cd5041d3f3b5e59761f36699167&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3% | [2.3%, 2.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.3% | [2.3%, 2.3%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b73e9a48aeeb44fa897ab820737dfd77c5076e01&end=9339f446a5302cd5041d3f3b5e59761f36699167&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 654.151s -> 654.227s (0.01%)

rylev commented:
@GuillaumeGomez This looks very much like a real regression (the spike in the graph below is the change caused by this PR):

<img width="354" alt="Screenshot 2023-08-05 at 14 52 38" src="https://github.com/rust-lang/rust/assets/1327285/09cc4105-245f-4be5-83e4-e5e4c6a8b381">

Looks like we might be spending much more time in`render_mod_item`. Any ideas?

GuillaumeGomez commented:
It's exactly as showed by the perf results. We considered the perf regression to be acceptable in regard to the fix.

lqd commented:
(it already was marked as `perf-regression-triaged` :)
